src/storage-client/src/controller.rs

```diff
@@ -2405,6 +2439,110 @@ where
         let json_state = serde_json::to_value(shard_state)?;
         Ok(json_state)
     }
+
+    async fn record_frontiers(&mut self, mut frontiers: BTreeMap<GlobalId, Self::Timestamp>) {
+        let old_rows = &mut self.frontiers_rows;
+        let mut updates = Vec::with_capacity(old_rows.len());
+
+        // Retract existing rows.
+        updates.extend(old_rows.drain(..).map(|row| (row, -1)));
+
+        let mut push_frontier = |object_id: GlobalId, time: T| {
+            let row = Row::pack_slice(&[
+                Datum::String(&object_id.to_string()),
+                Datum::MzTimestamp(mz_repr::Timestamp::from_bytes(time.encode())),
+            ]);
+            updates.push((row.clone(), 1));
+            old_rows.push(row);
+        };
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(collection_id);
+            if let Some(time) = collection.write_frontier.clone().into_option() {
+                push_frontier(*collection_id, time);
+            }
+        }
+        for (export_id, export) in &self.state.exports {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(export_id);
+            if let Some(time) = export.write_frontier.clone().into_option() {
+                push_frontier(*export_id, time);
+            }
+        }
+
+        // Add the provided frontiers.
+        for (object_id, time) in frontiers {
+            push_frontier(object_id, time);
+        }
+
+        // Cancel out superfluous retractions.
+        consolidate(&mut updates);
```

Comment:
Some of this logic could be removed if we stored the pristine previous `BTreeMap<GlobalId, Self::Timestamp>` in our state instead of a vec of rows which loses the key.

```rust
// Augment `frontiers` from local state here

// Replace previous `frontiers` and calculate consolidated diff
let mut updates = vec![];
for (id, new_time) in &self.reported_frontiers {
    match old_frontiers.remove(id) {
        Some(old_time) if old_time != new_time => {
            updates.push((encode_row(id, new_time), 1));
            updates.push((encode_row(id, old_time), -1));
        },
        None => updates.push((encode_row(id, new_time), 1)),
        _ => {},
    }
}
for (id, old_time) in old_frontiers {
    updates.push((encode_row(id, old_time), -1));
}
```